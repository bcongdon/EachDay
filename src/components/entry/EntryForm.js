import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Field, reduxForm } from 'redux-form'
import { PropTypes } from 'prop-types'
import { Button, Form } from 'semantic-ui-react'
import SemanticReduxFormField from '../form/SemanticReduxFormField'
import RatingFormField from '../form/RatingFormField'
import DatePickerFormField from '../form/DatePickerFormField'
import ErrorMessage from '../ErrorMessage'
import { closeEntryModal, createEntry, editEntry } from '../../actions'
import moment from 'moment'

const form = reduxForm({
  form: 'createEntry',
  validate
})

function validate (formProps) {
  const errors = {}

  if (!formProps.date || !moment(formProps.date).isValid()) {
    errors.date = 'Please enter a date'
  }

  return errors
}

class EntryForm extends Component {
  static propTypes = {
    handleSubmit: PropTypes.func.isRequired,
    isNewEntry: PropTypes.bool.isRequired,
    createEntry: PropTypes.func.isRequired,
    closeEntryModal: PropTypes.func.isRequired,
    editEntry: PropTypes.func.isRequired,
    errorMessage: PropTypes.string
  }

  constructor (props) {
    super(props)
    this.handleFormSubmit = this.handleFormSubmit.bind(this)
  }

  handleFormSubmit (formProps) {
    if (this.props.isNewEntry) {
      this.props.createEntry(formProps)
    } else {
      this.props.editEntry(formProps)
    }
    this.props.closeEntryModal()
  }

  renderAlert () {
    if (!this.props.errorMessage) {
      return null
    }

    return (
      <ErrorMessage message={this.props.errorMessage} />
    )
  }

  render () {
    return (
      <Form warning error onSubmit={this.props.handleSubmit(this.handleFormSubmit)}>
        {this.renderAlert()}
        <Form.Group>
          <Form.Field>
            <label>Rating</label>
            <Field
              component={RatingFormField} clearable name='rating'
              />
          </Form.Field>
          <Form.Field style={{marginLeft: 'auto'}}>
            <label>Date</label>
            <Field
              component={DatePickerFormField}
              name='date'
              />
          </Form.Field>
        </Form.Group>
        <Form.Field>
          <label>Notes</label>
          <Field
            component={SemanticReduxFormField}
            as={Form.TextArea}
            name='notes'
            placeholder={'How\'d the day go?'} />
        </Form.Field>
        <Button type='submit' primary>Save</Button>
      </Form>
    )
  }
}

function mapStateToProps (state) {
  return {
    errorMessage: state.entry.error
  }
}

export default connect(mapStateToProps, { createEntry, closeEntryModal, editEntry })(form(EntryForm))
