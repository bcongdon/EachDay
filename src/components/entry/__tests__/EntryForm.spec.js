import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { EntryForm } from '../EntryForm'
import { Button } from 'semantic-ui-react'

jest.mock('universal-cookie')

const handleSubmit = jest.fn()
const createEntry = jest.fn()
const closeEntryModal = jest.fn()
const editEntry = jest.fn()

describe('entry component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should render correctly with blank data', () => {
    var form = shallow(
      <EntryForm
        handleSubmit={handleSubmit}
        isNewEntry
        createEntry={createEntry}
        closeEntryModal={closeEntryModal}
        editEntry={editEntry}
        errorMessage=''
        />
    )
    expect(toJson(form)).toMatchSnapshot()
  })

  it('should render correctly with an error message', () => {
    var form = shallow(
      <EntryForm
        handleSubmit={handleSubmit}
        isNewEntry
        createEntry={createEntry}
        closeEntryModal={closeEntryModal}
        editEntry={editEntry}
        errorMessage='Something went wrong'
        />
    )
    expect(toJson(form)).toMatchSnapshot()
  })

  it('should call handleSubmit on submit', () => {
    var form = shallow(
      <EntryForm
        handleSubmit={handleSubmit}
        isNewEntry
        createEntry={createEntry}
        closeEntryModal={closeEntryModal}
        editEntry={editEntry}
        errorMessage=''
        />
    )
    form.find(Button).simulate('click')
    expect(handleSubmit).toBeCalled()
  })

  it('should call createEntry in handleFormSubmit if entry is new', () => {
    var form = shallow(
      <EntryForm
        handleSubmit={handleSubmit}
        isNewEntry
        createEntry={createEntry}
        closeEntryModal={closeEntryModal}
        editEntry={editEntry}
        errorMessage=''
        />
    )
    const formProps = {foo: 'bar'}
    form.instance().handleFormSubmit(formProps)
    expect(createEntry).toBeCalled()
    expect(createEntry.mock.calls[0][0]).toEqual(formProps)
    expect(editEntry).not.toBeCalled()
  })

  it('should call editEntry in handleFormSubmit if entry already exists', () => {
    var form = shallow(
      <EntryForm
        handleSubmit={handleSubmit}
        isNewEntry={false}
        createEntry={createEntry}
        closeEntryModal={closeEntryModal}
        editEntry={editEntry}
        errorMessage=''
        />
    )
    const formProps = {foo: 'bar'}
    form.instance().handleFormSubmit(formProps)
    expect(editEntry).toBeCalled()
    expect(editEntry.mock.calls[0][0]).toEqual(formProps)
    expect(createEntry).not.toBeCalled()
  })
})
