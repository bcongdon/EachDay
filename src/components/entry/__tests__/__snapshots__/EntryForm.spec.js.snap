// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entry component should render correctly with an error message 1`] = `
<Form
  as="form"
  error={true}
  warning={true}
>
  <ErrorMessage
    message="Something went wrong"
  />
  <FormGroup>
    <FormField>
      <label>
        Rating
      </label>
      <Field
        clearable={true}
        component={[Function]}
        name="rating"
      />
    </FormField>
    <FormField
      style={
        Object {
          "marginLeft": "auto",
        }
      }
    >
      <label>
        Date
      </label>
      <Field
        component={[Function]}
        name="date"
      />
    </FormField>
  </FormGroup>
  <FormField>
    <label>
      Notes
    </label>
    <Field
      as={[Function]}
      autoFocus={true}
      component={[Function]}
      name="notes"
      placeholder="How'd the day go?"
    />
  </FormField>
  <Button
    as="button"
    primary={true}
    type="submit"
  >
    Save
  </Button>
</Form>
`;

exports[`entry component should render correctly with blank data 1`] = `
<Form
  as="form"
  error={true}
  warning={true}
>
  <FormGroup>
    <FormField>
      <label>
        Rating
      </label>
      <Field
        clearable={true}
        component={[Function]}
        name="rating"
      />
    </FormField>
    <FormField
      style={
        Object {
          "marginLeft": "auto",
        }
      }
    >
      <label>
        Date
      </label>
      <Field
        component={[Function]}
        name="date"
      />
    </FormField>
  </FormGroup>
  <FormField>
    <label>
      Notes
    </label>
    <Field
      as={[Function]}
      autoFocus={true}
      component={[Function]}
      name="notes"
      placeholder="How'd the day go?"
    />
  </FormField>
  <Button
    as="button"
    primary={true}
    type="submit"
  >
    Save
  </Button>
</Form>
`;
